core/src/main/java/com/opensymphony/xwork2/util/fs/StrutsJarURLConnection.java

```diff
@@ -20,34 +20,89 @@
 
 import java.io.IOException;
 import java.io.InputStream;
+import java.io.UnsupportedEncodingException;
 import java.net.JarURLConnection;
 import java.net.MalformedURLException;
 import java.net.URL;
 import java.net.URLConnection;
+import java.net.URLDecoder;
 import java.nio.file.Files;
 import java.nio.file.Path;
 import java.nio.file.StandardCopyOption;
 import java.security.AccessController;
 import java.security.PrivilegedActionException;
 import java.security.PrivilegedExceptionAction;
+import java.util.jar.JarEntry;
 import java.util.jar.JarFile;
 
 /**
- * WW-4901 Decouples from underlying implementation of {@link URL#openConnection()}
+ * WW-4901 If was needed, decouples from underlying implementation of {@link URL#openConnection()}
  * e.g. from IBM WebSphere com.ibm.ws.classloader.Handler$ClassLoaderURLConnection
+ * WW-4920 Also decouples from and fixes {@link JarURLConnection#parseSpecs(URL)} if was needed
+ * e.g. from Oracle WebLogic which may report jar urls like "zip:C:/web-app-lib-path/some-jar.jar"
+ * but {@link JarURLConnection#parseSpecs(URL)} breaks on such urls
+ * While {@link JarURLConnection#parseSpecs(URL)} is private, then we had to extend {@link URLConnection} instead
  * @since 2.5.15
  */
-class StrutsJarURLConnection extends JarURLConnection {
+class StrutsJarURLConnection extends URLConnection {
+    private static final String FILE_URL_PREFIX = "file:";
+
+    private JarURLConnection jarURLConnection;
+
     private JarFile jarFile;
+    private String entryName;
+    private URL jarFileURL;
 
-    private StrutsJarURLConnection(URL url) throws MalformedURLException {
+    private StrutsJarURLConnection(URL url) throws IOException {
         super(url);
+
+        URLConnection conn = this.url.openConnection();
+        if (conn instanceof JarURLConnection) {//decoupling is not needed?
+            jarURLConnection = (JarURLConnection) conn;
+        } else {
+            try {
+                conn.getInputStream().close();
+            } catch (IOException ignored) {
+            }
+            parseSpecs(url);
+        }
     }
 
-    @Override
-    public JarFile getJarFile() throws IOException {
-        connect();
-        return jarFile;
+    /**
+    * A fixed copy of {@link JarURLConnection#parseSpecs(URL)}
+    */
+    private void parseSpecs(URL url) throws MalformedURLException, UnsupportedEncodingException {
```

Comment:
No unit tests for that. Or does `JarEntryRevisionTest` covers all?

Reply:
Yes it covers all according to [cover all report for this file](https://coveralls.io/builds/15608083/source?filename=core%2Fsrc%2Fmain%2Fjava%2Fcom%2Fopensymphony%2Fxwork2%2Futil%2Ffs%2FStrutsJarURLConnection.java#L74). But if you still prefer, I can write a sepearete test case named StrutsJarURLConnectionTest.java. Thanks!

Reply:
We can do it after the merge, but it would be great to have dedicated tests.

Reply:
I tried but it seems JUnit does not allow calling private methods. Then I researched and found [this answer](https://softwareengineering.stackexchange.com/questions/100959/how-do-you-unit-test-private-methods):

> You generally don't unit test private methods directly. Since they are private, consider them an implementation detail. Nobody is ever going to call one of them and expect it to work a particular way.